app/src/main/java/org/kiwix/kiwixmobile/local_file_transfer/WifiDirectManager.kt

```diff
@@ -0,0 +1,424 @@
+
+      override fun onSuccess() {
+        Log.d(TAG, "Disconnect successful")
+        closeChannel()
+      }
+    })
+  }
+
+  private fun closeChannel() {
+    if (VERSION.SDK_INT >= VERSION_CODES.O_MR1) {
+      channel!!.close()
+    }
+  }
+
+  fun getErrorMessage(reason: Int): String {
+    return when (reason) {
+      WifiP2pManager.ERROR -> "Internal error"
+      WifiP2pManager.BUSY -> "Framework busy, unable to service request"
+      WifiP2pManager.P2P_UNSUPPORTED -> "P2P unsupported on this device"
+      else -> "Unknown error code - $reason"
+    }
+  }
+
+  fun displayToast(
+    stringResourceId: Int,
+    templateValue: String,
+    duration: Int
+  ) {
+    LocalFileTransferActivity.showToast(
+      activity,
+      activity.getString(stringResourceId, templateValue),
+      duration
+    )
+  }
+
+  fun displayToast(stringResourceId: Int, duration: Int) {
+    LocalFileTransferActivity.showToast(activity, stringResourceId, duration)
+  }
+
+  fun onFileTransferAsyncTaskComplete(wereAllFilesTransferred: Boolean) {
+    if (wereAllFilesTransferred) {
+      displayToast(string.file_transfer_complete, Toast.LENGTH_LONG)
+    } else {
+      displayToast(string.error_during_transfer, Toast.LENGTH_LONG)
+    }
+    callbacks.onFileTransferComplete()
+  }
+
+  interface Callbacks {
+    fun onUserDeviceDetailsAvailable(userDevice: WifiP2pDevice?)
+    fun onConnectionToPeersLost()
+    fun updateListOfAvailablePeers(peers: WifiP2pDeviceList)
+    fun onFilesForTransferAvailable(filesForTransfer: ArrayList<FileItem>)
+    fun onFileStatusChanged(itemIndex: Int)
+    fun onFileTransferComplete()
+  }
+
+  companion object {
+    private const val TAG = "WifiDirectManager"
+    @JvmField var FILE_TRANSFER_PORT = 8008
+    @JvmStatic @Throws(IOException::class) fun copyToOutputStream(
+      inputStream: InputStream,
+      outputStream: OutputStream
+    ) {
+      val bufferForBytes = ByteArray(1024)
+      var bytesRead: Int
+      Log.d(TAG, "Copying to OutputStream...")
+      while (inputStream.read(bufferForBytes).also { bytesRead = it } != -1) {
+        outputStream.write(bufferForBytes, 0, bytesRead)
+      }
+      outputStream.close()
+      inputStream.close()
+      Log.d(LocalFileTransferActivity.TAG, "Both streams closed")
+    }
+
+    @JvmStatic fun getDeviceStatus(status: Int): String {
+      if (BuildConfig.DEBUG) Log.d(
+        TAG,
+        "Peer Status: $status"
+      )
+      return when (status) {
+        WifiP2pDevice.AVAILABLE -> "Available"
+        WifiP2pDevice.INVITED -> "Invited"
+        WifiP2pDevice.CONNECTED -> "Connected"
+        WifiP2pDevice.FAILED -> "Failed"
+        WifiP2pDevice.UNAVAILABLE -> "Unavailable"
+        else -> "Unknown"
+      }
+    }
+
+    @JvmStatic fun getFileName(fileUri: Uri): String {
+      val fileUriString = "$fileUri"
+      // Returns text after location of last slash in the file path
+      return fileUriString.substring(fileUriString.lastIndexOf('/') + 1)
```

Comment:
kotlin has a better method than this, finding that method is also left as an exercise for yourself